## Disciplina: MO911
## Tarefa 3 - Testes Estatísticos 

- Aluno: Rubens de Castro Pereira - RA217146
- Professor: Sandro Rigo 
- Data: 06/12/202

### Import libraries

In [181]:
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import scipy.stats as stats
from math import sqrt

### Loading datasets

O arquivo bp.csv contém dados sobre a pressão sanguínea (bp) e se a pessoa é ou não diabética (type).

In [182]:
df_bp = pd.read_csv('./datasets/bp - bp.csv')
df_bp

bp type
0    68   No
1    70  Yes
2    82   No
3    76   No
4    60   No
..   ..  ...
195  58   No
196  68  Yes
197  70   No
198  58   No
199  62  Yes

[200 rows x 2 columns]

***
### Questão 01

O arquivo bp.csv contem dados sobre a pressão sanguínea (bp) e se a pessoa é ou não diabética (type) (diabetico é type = Yes).

Use os seguintes testes e responda os itens abaixo:

- Teste t

- Wilcoxon rank sum

1. Reporte o p-valor da comparação entre a pressão sangüínea de pessoas diabéticas ou não.

2. Qual tem a maior pressão média? E maior pressão mediana?

3. A diferença é significativa?

Discuta a diferença dos p-valores. Quais dos 2 testes você acha mais apropriado nesse caso?

#### Preparing data

In [183]:
# preparing data 
diabetic_persons = df_bp.loc[df_bp['type'] == 'Yes']['bp']
# print(diabetic_persons)
diabetic_persons
print(f"Quantidade de pessoas diabéticas: {len(diabetic_persons)}")


non_diabetic_persons = df_bp.loc[df_bp['type'] == 'No']['bp']
# print(non_diabetic_persons)
non_diabetic_persons
print(f"Quantidade de pessoas não diabéticas: {len(non_diabetic_persons)}")


Quantidade de pessoas diabéticas: 68
Quantidade de pessoas não diabéticas: 132


#### Test t de Student

In [184]:
# calculating test t
print("Test t de Student")
test_t_stat, test_t_p_value = stats.ttest_ind(diabetic_persons, non_diabetic_persons)
print('stat=%.5f, p_value=%.5f' % (test_t_stat, test_t_p_value))

Test t de Student
stat=3.00150, p_value=0.00303


#### Test Wilcoxon rank sum

In [185]:
print("Test - Wilcoxon rank sum")
test_wilcoxon_stat, test_wilcoxon_p_value = stats.ranksums(diabetic_persons, non_diabetic_persons)
print('stat=%.5f, p_value=%.5f' % (test_wilcoxon_stat, test_wilcoxon_p_value))

Test - Wilcoxon rank sum
stat=3.04580, p_value=0.00232


#### Média e Mediana 

In [186]:
print(f"Diabéticos:")

diabetic_persons_mean = diabetic_persons.mean()
print(f"  Pressão Média ..: {diabetic_persons_mean}")

diabetic_persons_median = diabetic_persons.median()
print(f"  Pressão Mediana : {diabetic_persons_median}")

print()

print(f"Não diabéticos:")

non_diabetic_persons_mean = non_diabetic_persons.mean()
print(f"  Pressão Média ..: {non_diabetic_persons_mean}")

non_diabetic_persons_median = non_diabetic_persons.median()
print(f"  Pressão Mediana : {non_diabetic_persons_median}")

Diabéticos:
  Pressão Média ..: 74.58823529411765
  Pressão Mediana : 76.0

Não diabéticos:
  Pressão Média ..: 69.54545454545455
  Pressão Mediana : 70.0


### Questão 01 - Resultados

1. Reporte o p-valor da comparação entre a pressão sangüínea de pessoas diabéticas ou não.
    - O test Student T-test produziu um p-valor de 0.00303
    - O test Wilcoxon rank sum produziu um p-valor de 0.00232
    <br>
    
2. Qual tem a maior pressão média? E maior pressão mediana?
  
    - O grupo de pessoas diabéticas possui os maiores valores em relação às pessoas não diabéticas:
        - Pressão média: 74.59
        - Pressão mediana: 76.0
    <br>

3. A diferença é significativa?
    - Adotando-se como valor de decisão *p < 0,05*, os testes Test t de Student e Wilcoxon rank sum resultaram em valores *p_value*  bem menores, logo a diferença é significativa ou estatisticamente significante.
    <br>

4. Discuta a diferença dos p-valores. Quais dos 2 testes você acha mais apropriado nesse caso?

    - A amostra de dados possui um total de 200 valores de pressão sangüínea, sendo 68 pessoas diabéticas e 132 pessoas não diabéticas. Devido ao tamanho da amostra, o teste paramétrico test t de Student é mais apropriado para esse caso. Maiores informações sobre recomendação de uso do test t de Student podem ser verificadas em https://www.datascienceblog.net/post/statistical_test/parametric_sample_size/.

***
### Questão 02

O arquivo paired.csv contém dados sobre duas medições: em Agosto e em Novembro em 13 lugares diferentes.

Use os seguintes testes e responda os itens abaixo:
- Test t pareado
- Wilcoxon signed rank.

1. Qual o mês com maior média e maior mediana?

2. A diferença é significativa?

3. Discuta a diferença dos p-valores. Quais dos 2 testes você acha mais apropriado nesse caso?

#### Loading datasets

O arquivo paired.csv contém dados sobre duas medições: em Agosto e em Novembro em 13 lugares diferentes.

In [187]:
df_paired = pd.read_csv('./datasets/paired - paired.csv')
df_paired

Clone  August  November
0     Balsam_Spire     8.1      11.2
1          Beaupre    10.0      16.3
2        Hazendans    16.5      15.3
3        Hoogvorst    13.6      15.6
4         Raspalje     9.5      10.5
5             Unal     8.3      15.5
6   Columbia_River    18.3      12.7
7    Fritzi_Pauley    13.3      11.1
8        Trichobel     7.9      19.9
9            Gaver     8.1      20.4
10          Gibecq     8.9      14.2
11           Primo    12.6      12.7
12       Wolterson    13.4      36.8

#### Preparing data

In [188]:
# preparing data 
paired_august = df_paired['August']
paired_august

0      8.1
1     10.0
2     16.5
3     13.6
4      9.5
5      8.3
6     18.3
7     13.3
8      7.9
9      8.1
10     8.9
11    12.6
12    13.4
Name: August, dtype: float64

In [189]:
paired_november = df_paired['November']
paired_november

0     11.2
1     16.3
2     15.3
3     15.6
4     10.5
5     15.5
6     12.7
7     11.1
8     19.9
9     20.4
10    14.2
11    12.7
12    36.8
Name: November, dtype: float64

In [190]:
print(f"Tamanho das amostras:")
print(f"  August ..: {len(paired_august)}")
print(f"  November : {len(paired_november)}")

Tamanho das amostras:
  August ..: 13
  November : 13


#### Teste t de Student pareado

In [191]:
print("Test t de Student pareado")
test_t_pareado_stat, test_t_pareado_p_value = stats.ttest_rel(paired_august, paired_november)
print('stat=%.6f, p_value=%.6f' % (test_t_pareado_stat, test_t_pareado_p_value))

Test t de Student pareado
stat=-2.308896, p_value=0.039555


#### Test Wilcoxon signed rank

In [192]:
print("Test Wilcoxon signed rank")
test_wilcoxon_signed_rank_stat, test_wilcoxon_signed_rank_p_value = stats.wilcoxon(paired_august, paired_november)
print('stat=%.6f, p_value=%.6f' % (test_wilcoxon_signed_rank_stat, test_wilcoxon_signed_rank_p_value))

Test Wilcoxon signed rank
stat=16.000000, p_value=0.039795


#### Diferença dos valores *p_value* dos testes

In [193]:
print('Diferença p_value = %.10f' % (test_wilcoxon_signed_rank_p_value - test_t_pareado_p_value))
print(f'Diferença p_value = {(test_wilcoxon_signed_rank_p_value - test_t_pareado_p_value):e}')

Diferença p_value = 0.0002396710
Diferença p_value = 2.396710e-04


#### Média e Mediana

In [194]:
print(f"August:")
paired_august_mean = paired_august.mean()
print(f"  Média ..: {paired_august_mean}")

paired_august_median = paired_august.median()
print(f"  Mediana : {paired_august_median}")

print(f"November:")

paired_november_mean = paired_november.mean()
print(f"  Média ..: {paired_november_mean}")

paired_november_median = paired_november.median()
print(f"  Mediana : {paired_november_median}")

August:
  Média ..: 11.423076923076925
  Mediana : 10.0
November:
  Média ..: 16.323076923076922
  Mediana : 15.3


### Questão 02 - Resultados

1. Qual o mês com maior média e maior mediana?
    - O mês com maiores valores foi *November*:
        - Média: 16.32
        - Mediana: 15.3
    <br>

2. A diferença é significativa?
    - Adotando-se como valor de decisão *p < 0,05*, os testes Test t de Student pareado e Wilcoxon signed rank produziram  valores *p_value* bem menores do que *p*, logo a diferença é significativa ou estatisticamente significante.        
    <br>

3. Discuta a diferença dos p-valores. Quais dos 2 testes você acha mais apropriado nesse caso?
    - Os valores obtidos dos testes são:
        - Test Student T-test pareado: 0.039555
        - Test Wilcoxon signed rank: 0.039795
        - Diferença entre os valores dos dois testes: 2.396710e-04 
        - Tamanho das amostras: August: 13, and November: 13
    <br>
    
    - A amostra dos dados é pequena, e portanto, testes não paramétricos são mais apropriados. Por isso, o teste de Wilcoxon signed rank é mais apropriado, pois não possível estabelecer uma distribuição dos dados gerados. Maiores informações sobre recomendação de uso do test Wilcoxon signed rank podem ser verificadas em https://www.datascienceblog.net/post/statistical_test/signed_wilcox_rank_test/.

***
### Questão 03

Use os dados (não pareados) do dataset bp.csv

Gere o intervalo de confiança para a pressão sangüínea dos pacientes com diabetes e sem diabetes (intervalo de 95% de confiança) usando o método de bootstrap

In [195]:
# calculating IC by boostrap method to diabetics 
diabetic_persons_seq = (diabetic_persons.values,)
bootstrap_ci_diabetics = stats.bootstrap(diabetic_persons_seq, np.mean, confidence_level=0.95, random_state=1, method='BCa')
print("Intervalo de confiança de 95% para o grupo de diabéticos")
print(bootstrap_ci_diabetics.confidence_interval)

print()

# calculating IC by boostrap method to non diabetics 
non_diabetic_persons_seq = (non_diabetic_persons.values,)
bootstrap_ci_non_diabetics = stats.bootstrap(non_diabetic_persons_seq, np.mean, confidence_level=0.95, random_state=1, method='BCa')
print("Intervalo de confiança de 95% para o grupo de não diabéticos")
print(bootstrap_ci_non_diabetics.confidence_interval)

Intervalo de confiança de 95% para o grupo de diabéticos
ConfidenceInterval(low=71.77941176470588, high=77.29411764705883)

Intervalo de confiança de 95% para o grupo de não diabéticos
ConfidenceInterval(low=67.6590909090909, high=71.47727272727273)


***
### Questão 04

Teste Visual

- Há intersecção entre os intervalos de confiança dos 2 conjuntos de dados acima (com diabetes e sem diabetes)? 
    - Não há interseção.


- Isso está de acordo com o teste de diferença significativa entre eles feito na questão 1?

    - O resultado está em acordo com os testes de diferença de significativa, já que os intervalos de confiança não apresentam interseção.
    - como não houve interseção, o -pvalue é menor que 0,05


***
### Questão 05

Calcule o Cohen D com o pooled standard deviation para a diferença entre a pressão sangüínea de pacientes com e sem diabetes.

In [196]:
 
# function to calculate Cohen's d for independent samples
def cohend(d1, d2):
	# calculate the size of samples
	n1, n2 = len(d1), len(d2)
	# calculate the variance of the samples
	s1, s2 = np.var(d1, ddof=1), np.var(d2, ddof=1)
	# calculate the pooled standard deviation
	s = sqrt(((n1 - 1) * s1 + (n2 - 1) * s2) / (n1 + n2 - 2))
	# calculate the means of the samples
	u1, u2 = np.mean(d1), np.mean(d2)
	# calculate the effect size
	return (u1 - u2) / s
 
# calculate cohen's d
difference = cohend(diabetic_persons, non_diabetic_persons)
print('Cohen D: %.6f' % difference)

Cohen D: 0.448034


*** 
### Questão 06

Com base nos testes, intervalos de confiança, etc realizados acima sobre o dataset bp.csv, qual a sua conclusão sobre pressão arterial em pacientes com ou sem diabetes?

- A pressãp arterial elevada em pacientes  do grupo dos diabéticos é maior do 

- o IC das médias não intersecciona e o o pvalor são diferentes, são menores que 0,05, e por isso podemos concluir que pessoas diaticas tem maior risco de ter presão arterial elevada.